packages/core/admin/server/src/services/homepage.ts

```diff
@@ -1,129 +1,218 @@
+      const fields = ['documentId', 'updatedAt'];
+
+      // Add fields required to get the status if D&P is enabled
+      const hasDraftAndPublish = contentTypes.hasDraftAndPublish(contentType);
+      if (hasDraftAndPublish) {
+        fields.push('publishedAt');
+      }
+
+      // Only add the main field if it's defined
+      if (configuration?.settings.mainField) {
+        fields.push(configuration.settings.mainField);
+      }
+
+      // Only add locale if it's localized
+      const isLocalized = (contentType.pluginOptions?.i18n as any)?.localized;
+      if (isLocalized) {
+        fields.push('locale');
+      }
+
+      return {
+        fields,
+        mainField: configuration!.settings.mainField,
+        contentType,
+        hasDraftAndPublish,
+        uid,
+      };
+    });
+  };
 
-          // Only add locale if it's localized
-          const isLocalized = (contentType.pluginOptions?.i18n as any)?.localized;
-          if (isLocalized) {
-            fields.push('locale');
+  const formatDocuments = (documents: Modules.Documents.AnyDocument[], meta: DocumentMeta) => {
+    return documents.map((document) => {
+      /**
+       * Save the main field value before deleting it so we can use the common
+       * title key instead across all content types. Use the delete operator instead of
+       * destructuring or lodash omit for better type inference.
+       */
+      const mainFieldValue = document[meta.mainField ?? 'documentId'];
+      delete document[meta.mainField];
+
+      return {
+        data: {
+          ...document,
+          ...(document.publishedAt && { publishedAt: new Date(document.publishedAt) }),
+          updatedAt: new Date(document.updatedAt),
+          title: mainFieldValue,
+        },
+        meta: {
+          model: meta.uid,
+          kind: meta.contentType.kind,
+          hasDraftAndPublish: meta.hasDraftAndPublish,
+        },
+      };
+    });
+  };
+
+  const addStatusToDocuments = async (
+    documents: {
+      data: RecentDocument[];
+      meta: {
+        model: RecentDocument['model'];
+        kind: RecentDocument['kind'];
+        hasDraftAndPublish: boolean;
+      };
+    }[]
+  ): Promise<RecentDocument[]> => {
+    return Promise.all(
+      documents.map(async (document) => {
+        /**
+         * Tries to query the other version of the document if draft and publish is enabled,
+         * so that we know when to give the "modified" status.
+         */
+        const { availableStatus } = await metadataService.getMetadata(
+          document.meta.model,
+          document.data,
+          {
+            availableStatus: document.meta.hasDraftAndPublish,
+            availableLocales: false,
           }
+        );
+        const status = metadataService.getStatus(document.data, availableStatus);
+
+        return {
+          ...document.data,
+          ...document.meta,
+          status,
+        } as unknown as RecentDocument;
```

Comment:
ℹ️ I just started getting frustrated. If you have ideas to improve the TS let me know.

Reply:
this makes me a bit sad, what was the error?

Reply:
The problem TypeScript isn't smart enough to know what's in the spread. So it thinks all the properties are missing.

Reply:
I see you removed the delete keyword to use destructuring, I think this is the issue I had without that delete

Reply:
I think 🤔 , idk I'm getting turned around with the 2D array. 

Here's the error:

```
Type '{ status: any; model: ContentType; kind: ContentTypeKind; hasDraftAndPublish: boolean; length: number; toString(): string; toLocaleString(): string; toLocaleString(locales: string | string[], options?: (NumberFormatOptions & DateTimeFormatOptions) | undefined): string; ... 37 more ...; [Symbol.unscopables]: { ...; }...' is not assignable to type 'RecentDocument[]'.
  Type '{ status: any; model: ContentType; kind: ContentTypeKind; hasDraftAndPublish: boolean; length: number; toString(): string; toLocaleString(): string; toLocaleString(locales: string | string[], options?: (NumberFormatOptions & DateTimeFormatOptions) | undefined): string; ... 37 more ...; [Symbol.unscopables]: { ...; }...' is missing the following properties from type 'RecentDocument': documentId, title, updatedAt
```

Reply:
> I see you removed the delete keyword to use destructuring, I think this is the issue I had without that delete

We do lose type inference and I will have to revert that change but it doesn't completely resolve the issue here. I find that so frustrating that we can't use better JavaScript features because TypeScript doesn't know what to do with them.

Reply:
nice, this is simpler and makes sense. testing now